### PR TITLE
Handle Color3 type

### PR DESCRIPTION
**Pull request by jeparlefrancais**
I just started to use remodel for a project and I find it pretty useful! 

I'm currently using it to build a model from a set of place files: the goal is to have a separate place for each world, and then read each of them with remodel and extract the necessary data.

I ended up reading the current Lighting properties of each place and creating a series of Value objects to include into the final model. So in order to serialize the Ambient property, for example, I needed that Color3 value.

I know there was not any issues related to that so I don't know if you had something else in mind, just let me know.

### CODE REVIEW

File: src/value.rs [hunk 1]
```diff
@@ -91,6 +93,13 @@ pub fn lua_to_rbxvalue(ty: RbxValueType, value: LuaValue<'_>) -> LuaResult<RbxVa
             value: value as i64,
         }),
 
+        (RbxValueType::Color3, LuaValue::UserData(ref user_data))
+            if user_data.is::<Color3Value>() =>
+        {
+            let color = &*user_data.borrow::<Color3Value>().unwrap();
+            Ok(color.into())
+        }
+
```
Comment (LPGhatguy) [inline src/value.rs]: Since `user_data.borrow::<T>()` will tell us if we couldn't do that borrow, I think we should drop the `if` guard on the pattern and do that check internally.

We might need to finagle it a bit to get it to compile, but something like this should work:

```rust
(RbxValueType::Color3, LuaValue::UserData(ref user_data)) => {
    let color = user_data.borrow::<Color3Value>()?;
    Ok(color.into())
}
```

Note that the error type returned by `borrow` is `rlua::Error` I think, which means we can use Rust's `?` operator!

This means we'll be duplicating the error return case, but I think that's okay. If it ends up being a problem, I like [hoisting the error type into a closure](https://github.com/rojo-rbx/rbx-dom/blob/412d996e55ceb261eaf75f541cb7bf41f2778eeb/rbx_binary/src/serializer.rs#L459-L470).

File: src/value.rs [hunk 2]
```diff
@@ -132,3 +141,33 @@ pub fn type_from_str(name: &str) -> Option<RbxValueType> {
...
+}
+
+impl Color3Value {
+    pub fn new(value: [f32; 3]) -> Self {
+        Self { value }
+    }
+
+    fn meta_to_string<'lua>(&self, context: Context<'lua>) -> rlua::Result<rlua::Value<'lua>> {
+        format!("{}, {}, {}", self.value[0], self.value[1], self.value[2]).to_lua(context)
+    }
```
Comment (LPGhatguy) [inline src/value.rs]: Let's port this code to a `Display` impl on the type! Then, when we implement the `__tostring` metamethod below, we can do it as `Ok(this.to_string())`.

File: src/value.rs [hunk 3]
```diff
@@ -132,3 +141,33 @@ pub fn type_from_str(name: &str) -> Option<RbxValueType> {
         _ => None,
     }
 }
+
+struct Color3Value {
```
Comment (LPGhatguy) [inline src/value.rs]: We should be able to derive `Debug`, `Clone`, and `Copy` on this type to make our lives easier elsewhere.

File: src/value.rs [hunk 4]
```diff
@@ -132,3 +141,33 @@ pub fn type_from_str(name: &str) -> Option<RbxValueType> {
...
+    }
+}
+
+impl From<&Color3Value> for RbxValue {
+    fn from(color: &Color3Value) -> RbxValue {
+        RbxValue::Color3 {
+            value: color.value.clone(),
+        }
+    }
+}
```
Comment (LPGhatguy) [inline src/value.rs]: If we make our type `Copy`, we can drop all the references and `clone()` calls, which makes this a little easier.

**Comment by LPGhatguy** [comment]
Another thing crossed my mind, we should try to handle cases where the property type is `Color3uint8` as well! It's mostly a serialization distinction, but some color properties can have components that exceed 1.0 while others cannot.

**Comment by jeparlefrancais** [comment]
I'd like to add some tests around the changes that I made, but I am not exactly sure how I should do them. First, I tried adding a script that would assert on the Value property of a Color3Value, but the problem I had was that there is no "Value" property when you create the instance:

```lua
local color3Value = Instance.new("Color3Value")

print(remodel.getRawProperty(color3Value, "Value")) -- this prints nil

remodel.setRawProperty(color3Value, "Value", "Color3", someColor3ReadFromSomewhere)

print(remodel.getRawProperty(color3Value, "Value")) -- this prints the color value now
```

I guess that is because the properties are like "lazy" loaded since it's weakly typed right? So another option could be a script like this, but I feel like if we could avoid to read a file it would be better:

```lua
local game = remodel.readPlaceFile("test-models/place-with-models.rbxlx")
local ambient = remodel.getRawProperty(game.Lighting, "Ambient")

assert(tostring(ambient) == "0, 0, 0")
```

Do you have any other ideas?

**Comment by LPGhatguy** [comment]
I guess we have a couple of options for testing this!

1. Implement support for Roblox-style property access with default values. This would make our tests the most natural, but might be kind of a difficult undertaking!
2. Create files in Roblox Studio that we can deserialize and measure from. I like this option, especially since the XML format is easy to audit the representation of. I would probably create more focused model files instead of poking at services in places, if we can.
3. Test reading + writing in the same go. For this test, we'd construct a `Color3Value`, set `Value`, then read `Value` back. This might make us fall victim to a symmetrical error, which I don't like, but the tests are self-contained at least.

My vote would be for option 2, I think. It's straightforward (don't need to do any more work) and lets us exercise a lot of corner cases. It has a downside that the exact behavior will depend on the serialization libraries pretty heavily.

**Comment by LPGhatguy** [comment]
Let's merge this and iterate on more types!
